target/linux/ramips/dts/ApexeraWiBot.dts

```diff
@@ -0,0 +1,169 @@
+/dts-v1/;
+
+#include "mt7620a.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "apexera,wibot", "ralink,mt7620a-soc";
+	model = "Apexera WiBot";
+
+	chosen {
+		bootargs = "console=ttyS1,57600";
+	};
+
+	aliases {
+		led-boot = &led_boot;
+		led-failsafe = &led_status;
+		led-running = &led_status;
+		led-upgrade = &led_status;
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		wan {
+			label = "wibot:blue:wan";
+			gpios = <&gpio1 5 GPIO_ACTIVE_HIGH>;
+		};
+
+		wifi {
+			label = "wibot:blue:wifi";
+			gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status: status {
+			label = "wibot:blue:status";
+			gpios = <&gpio1 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_boot: boot {
+			label = "wibot:blue:boot";
+			gpios = <&gpio1 11 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		reset {
+			label = "wibot:blue:reset";
+			gpios = <&gpio2 23 GPIO_ACTIVE_HIGH>;
+		};
+
+		ctrl3g {
+			label = "wibot:blue:ctrl3g";
+			gpios = <&gpio2 21 GPIO_ACTIVE_HIGH>;
+		};
+
+		usb0 {
+			label = "wibot:blue:usb0";
+			gpios = <&gpio2 20 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		usb1 {
+			label = "wibot:blue:usb1";
+			gpios = <&gpio2 22 GPIO_ACTIVE_HIGH>;
+		};
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&gpio2 {
+	status = "okay";
+};
+
+&uart {
+	status = "okay";
+};
+
+&i2c {
+	status = "okay";
+
+	pcf8563: rtc@51 {
+		status = "okay";
+		compatible = "nxp,pcf8563";
+		reg = <0x51>;
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0xfb0000>;
+			};
+		};
+	};
+};
+
+&sdhci {
+	status = "okay";
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
+
+&pcie {
+	status = "okay";
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x10>;
+	ralink,port-map = "llllw";
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "rgmii1", "rgmii2";
+			ralink,function = "gpio";
+		};
+
+		uartf_gpio {
```

Comment:
This pinmux doesn't make much sense to me. You switch the full uart to uart light, to use &gpio0 11 to 14 as gpios. But I can't see them used in the dts.

Any explanation why you are doing it?

Reply:
We have a stm32f030fep6 chip with this box, which manages the charging state of battery. and the stm32 chip communicates with mt7620 with i2c. so we use this uart as the i2c interface. it seems amazing and weird.

Reply:
And for the detailed protocol of this charging stm32 chip, we have to dig in.

Reply:
I still don't get it. If the stm32 communicates with the mt7620 via i2c, what are these GPIOs used for?

Reply:
they are left, which could be used as gpio-i2c for communicating.

Reply:
Got it, the stm32 is not connected to the mt7620 i2c bus as I assumed. Instead you are doing bit banging i2c via gpio.

Reply:
yes.

Reply:
would you please add a node for the [i2c-gpio driver](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/i2c/i2c-gpio.txt).

You will find some more examples by grepping for "i2c-gpio" in the ramips dts files.

I am afraid that the pinmux will be dropped at some point, as it looks like there are no users. Similar to my initial assumption.

Reply:
are you working on this? this one is a must have.

Reply:
I have updated the device description, I have made some mistakes. 

                     |---------i2c-------->|
mt7620a------|                            |---uart(i2c)---->stm32f0
                      |------uart1-------->|

This is how the mt7620a and stm32f0 connected. the functions as below:

1. disable i2c, we can use uart1 to upgrade the firmware of stm32f0
2. disable uart1, we can use i2c to manage the charging state with stm32f0